From 627372902f4552be486355fbdddfcf1f41d6b32d Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Sun, 07 Apr 2019 18:54:18 +0300
Subject: [PATCH] [SW-0] preboot: replace io accessors with local implementation

Replace U-boot readb/readl/writeb/writel with local implementation
taken from fw_loader.
It removes part of the coupling with u-boot project.

Change-Id: Iac4bd5f5ec7d900a8880e0c3bd92da1ebd5f2e87
---

diff --git a/common/preboot.c b/common/preboot.c
index be7fce6..9244add 100644
--- a/common/preboot.c
+++ b/common/preboot.c
@@ -35,17 +35,18 @@
 	bool try_new_image = false;
 	char *spl_image = "candidate";
 
-	modifiable_image_pending = readl(mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
+	modifiable_image_pending =
+			regmap_readl(mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
 	if (modifiable_image_pending == FLOPS_UPD_PENDING) {
 		next_image_modifiable =
 			(*(unsigned int *)(HL_SPI_BASE + 0x60004) & 0xff) - '1';
 		if (next_image_modifiable < 2) {
 			next_image = next_image_modifiable;
-			writel((next_image + FLOPS_UPD_NXT_IMG1),
+			regmap_writel((next_image + FLOPS_UPD_NXT_IMG1),
 					mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
 			try_new_image = true;
 		} else
-			writel(FLOPS_UPD_NOOP,
+			regmap_writel(FLOPS_UPD_NOOP,
 					mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3);
 	}
 
@@ -71,7 +72,7 @@
 	preboot_puts(")\n");
 
 	*spl_off = image_base_offset + spl_offset;
-	writel((image_base_offset + uboot_offset),
+	regmap_writel((image_base_offset + uboot_offset),
 			mmPSOC_GLOBAL_CONF_SCRATCHPAD_30);
 }
 
@@ -87,7 +88,7 @@
 
 static void preboot_pcie_init(void)
 {
-	if (readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS) & SKIP_PCIE_INIT) {
+	if (regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS) & SKIP_PCIE_INIT) {
 		preboot_puts("PCIe initialization is skipped!\n");
 		return;
 	}
@@ -103,7 +104,7 @@
 static void preboot_version(void)
 {
 	u64 offset = (u64) preboot_btl_ver_str;
-	u32 btlver = readl(mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
+	u32 btlver = regmap_readl(mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
 	int i, btl_ver_sz;
 
 	preboot_puts("\n");
@@ -131,7 +132,7 @@
 			break;
 	}
 
-	writel((u32)(offset & ~CONFIG_SYS_SDRAM_BASE),
+	regmap_writel((u32)(offset & ~CONFIG_SYS_SDRAM_BASE),
 			mmPSOC_GLOBAL_CONF_SCRATCHPAD_28);
 }
 #else	/* TARGET_IS_TESTER */
@@ -142,7 +143,7 @@
 {
 	void (*spl_start)(void) = (void *)CONFIG_SPL_TEXT_BASE;
 
-	writel(CPU_BOOT_STATUS_IN_PREBOOT, get_cpu_boot_status_reg());
+	regmap_writel(CPU_BOOT_STATUS_IN_PREBOOT, get_cpu_boot_status_reg());
 
 	preboot_version();
 	preboot_pcie_init();
diff --git a/common/preboot.h b/common/preboot.h
index 47da923..b5b17d0 100644
--- a/common/preboot.h
+++ b/common/preboot.h
@@ -8,6 +8,30 @@
 #ifndef COMMON_PREBOOT_H_
 #define COMMON_PREBOOT_H_
 
+static inline u8 regmap_readb(u64 offset)
+{
+	u8 reg = *((volatile u8 *)offset);
+
+	return reg;
+}
+
+static inline void regmap_writeb(u8 val, u64 offset)
+{
+	*((volatile u8 *)offset) = val;
+}
+
+static inline u32 regmap_readl(u64 offset)
+{
+	u32 reg = *((volatile u32 *)offset);
+
+	return reg;
+}
+
+static inline void regmap_writel(u32 val, u64 offset)
+{
+	*((volatile u32 *)offset) = val;
+}
+
 u64 get_cpu_boot_status_reg(void);
 
 #endif /* COMMON_PREBOOT_H_ */
diff --git a/common/spi.c b/common/spi.c
index d4408d3..323bac0 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -11,6 +11,7 @@
 #include <linux/bitops.h>
 #include <mach/hl_regs.h>
 
+#include "preboot.h"
 #include "spi.h"
 #include "uart.h"
 
@@ -53,13 +54,13 @@
 {
 	unsigned int reg, mask = BIT(53 % 16);
 
-	reg  = readl(mmPSOC_GPIO3_DATAOUT);
+	reg  = regmap_readl(mmPSOC_GPIO3_DATAOUT);
 	reg  &= ~mask;
-	writel(reg,  mmPSOC_GPIO3_DATAOUT);
+	regmap_writel(reg,  mmPSOC_GPIO3_DATAOUT);
 
-	reg = readl(mmPSOC_GPIO3_OUTENSET);
+	reg = regmap_readl(mmPSOC_GPIO3_OUTENSET);
 	reg |=  mask;
-	writel(reg, mmPSOC_GPIO3_OUTENSET);
+	regmap_writel(reg, mmPSOC_GPIO3_OUTENSET);
 }
 
 static void spi_claim_bus(void)
@@ -69,7 +70,7 @@
 	if (!reg)
 		reg = get_cs_pad_sel();
 
-	writel(1, reg);
+	regmap_writel(1, reg);
 }
 
 static void spi_release_bus(void)
@@ -79,19 +80,19 @@
 	if (!reg)
 		reg = get_cs_pad_sel();
 
-	writel(0, reg);
+	regmap_writel(0, reg);
 }
 
 static inline void spi_dev_update_frame_size(unsigned int len)
 {
-	writel(0, mmPSOC_SPI_SSIENR);
-	writel(frame_size_bits((len * 8) - 1), mmPSOC_SPI_CTRLR0);
-	writel(1, mmPSOC_SPI_SSIENR);
+	regmap_writel(0, mmPSOC_SPI_SSIENR);
+	regmap_writel(frame_size_bits((len * 8) - 1), mmPSOC_SPI_CTRLR0);
+	regmap_writel(1, mmPSOC_SPI_SSIENR);
 }
 
 static inline int spi_fifo_is_empty(unsigned long level_reg)
 {
-	return !readl(level_reg);
+	return !regmap_readl(level_reg);
 }
 
 static void spi_dev_dword_cmd(unsigned int cmd, unsigned int *val)
@@ -100,11 +101,11 @@
 
 	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
 		;
-	writel(cmd, mmPSOC_SPI_DR0);
+	regmap_writel(cmd, mmPSOC_SPI_DR0);
 
 	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
 		;
-	tmp = readl(mmPSOC_SPI_DR0);
+	tmp = regmap_readl(mmPSOC_SPI_DR0);
 
 	if (val)
 		*val = tmp;
@@ -118,11 +119,11 @@
 
 	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
 		;
-	writeb(cmd, mmPSOC_SPI_DR0);
+	regmap_writeb(cmd, mmPSOC_SPI_DR0);
 
 	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
 		;
-	tmp = readb(mmPSOC_SPI_DR0);
+	tmp = regmap_readb(mmPSOC_SPI_DR0);
 
 	if (val)
 		*val = tmp;
@@ -151,20 +152,20 @@
 
 	while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
 		;
-	writeb(cmd, mmPSOC_SPI_DR0);
+	regmap_writeb(cmd, mmPSOC_SPI_DR0);
 	for (i = 0; i < size; i++) {
 		while (!spi_fifo_is_empty(mmPSOC_SPI_TXFLR))
 			;
-		writeb(val[i], mmPSOC_SPI_DR0);
+		regmap_writeb(val[i], mmPSOC_SPI_DR0);
 	}
 
 	while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
 		;
-	readb(mmPSOC_SPI_DR0);
+	regmap_readb(mmPSOC_SPI_DR0);
 	for (i = 0; i < size; i++) {
 		while (spi_fifo_is_empty(mmPSOC_SPI_RXFLR))
 			;
-		readb(mmPSOC_SPI_DR0);
+		regmap_readb(mmPSOC_SPI_DR0);
 	}
 
 	spi_release_bus();
@@ -213,13 +214,13 @@
 static void spi_init(void)
 {
 	spi_ssi_cs0_init();
-	writel(0x00000000, mmPSOC_SPI_SSIENR);
-	writel(0x00000002, mmPSOC_SPI_BAUDR);	/* Min divisor */
-	writel(0x00000000, mmPSOC_SPI_RXFTLR);
-	writel(0x00000000, mmPSOC_SPI_TXFTLR);
-	writel(0x0000003F, mmPSOC_SPI_IMR);	/* Disable Interrupts */
-	writel(0x00000001, mmPSOC_SPI_SER);
-	writel(0x00000001, mmPSOC_SPI_SSIENR);
+	regmap_writel(0x00000000, mmPSOC_SPI_SSIENR);
+	regmap_writel(0x00000002, mmPSOC_SPI_BAUDR);	/* Min divisor */
+	regmap_writel(0x00000000, mmPSOC_SPI_RXFTLR);
+	regmap_writel(0x00000000, mmPSOC_SPI_TXFTLR);
+	regmap_writel(0x0000003F, mmPSOC_SPI_IMR);	/* Disable Interrupts */
+	regmap_writel(0x00000001, mmPSOC_SPI_SER);
+	regmap_writel(0x00000001, mmPSOC_SPI_SSIENR);
 }
 
 void spi_copy_spl(unsigned int spl_offset)
diff --git a/common/uart.c b/common/uart.c
index 09e9c7a..0e89a8c 100644
--- a/common/uart.c
+++ b/common/uart.c
@@ -7,6 +7,7 @@
 
 #include <linux/io.h>
 #include <mach/hl_regs.h>
+#include "preboot.h"
 
 #define mmPSOC_UART_0_DATA		(HL_CFG_BASE + 0xC45000)
 #define mmPSOC_UART_0_STATE		(HL_CFG_BASE + 0xC45004)
@@ -18,9 +19,9 @@
 	int i;
 
 	for (i = 0; str[i] && (i < 128); i++) {
-		while (readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL)
+		while (regmap_readl(mmPSOC_UART_0_STATE) & UART_SR_TXFULL)
 			;
-		writel(str[i], mmPSOC_UART_0_DATA);
+		regmap_writel(str[i], mmPSOC_UART_0_DATA);
 		preboot_putc_virt_uart(str[i]);
 	}
 }
diff --git a/gaudi/gaudi.c b/gaudi/gaudi.c
index cabee12..9fe7cbc 100644
--- a/gaudi/gaudi.c
+++ b/gaudi/gaudi.c
@@ -9,13 +9,15 @@
 #include <mach/hl_regs.h>
 #include <mach/../../gaudi/include/mach/gaudi_regs.h>
 
+#include "preboot.h"
+
 void preboot_pcie_wait_for_link(void)
 {
 	while (1) {
 		unsigned int smlh, rdlh;
 
-		smlh = readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = readl(mmPCIE_AUX_RDLH_LINK_UP);
+		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
 		if (smlh && rdlh)
 			break;
 	}
diff --git a/goya/goya.c b/goya/goya.c
index 9cd60296..363ad29 100644
--- a/goya/goya.c
+++ b/goya/goya.c
@@ -9,6 +9,7 @@
 #include <linux/io.h>
 #include <mach/hl_regs.h>
 #include <mach/../../goya/include/mach/goya_regs.h>
+#include "preboot.h"
 
 #define BTL_COUNTER_MAX		BIT(22)
 #define SYS_REF_CLOCK		50000000
@@ -19,7 +20,7 @@
 	unsigned long count = (timeout_ns * SYS_REF_CLOCK) / 1000000000UL;
 
 	while (count--) {
-		val = readl(reg);
+		val = regmap_readl(reg);
 		if (val & 1)
 			return 0;
 	}
@@ -35,12 +36,12 @@
 	 * FSM will set ext_ld_done after 32KB but why wait
 	 * if we can already move on
 	 */
-	while ((readl(mmPCIE_PHY_CR_PARA_CMD) & 0xFFFF) <= 0xD780)
+	while ((regmap_readl(mmPCIE_PHY_CR_PARA_CMD) & 0xFFFF) <= 0xD780)
 		;
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0);
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_1);
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_2);
-	writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3);
+	regmap_writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_0);
+	regmap_writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_1);
+	regmap_writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_2);
+	regmap_writel(1, mmPCIE_PHY_PHYX_SRAM_EXT_LD_DONE_3);
 }
 
 static void preboot_wait_for_stable_pclk(void)
@@ -74,26 +75,26 @@
 	preboot_wait_for_ext_ld_done();
 
 	/* Wait for PRSTn */
-	while (!readl(mmPSOC_GLOBAL_CONF_PRSTN))
+	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
 		;
 
 	/* Enable FW FSM after PRSTn */
-	writel(1, mmPSOC_GLOBAL_CONF_PCI_FW_FSM);
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCI_FW_FSM);
 
 	preboot_wait_for_stable_pclk();
 
 	while (1) {
 		unsigned int prstn, ltssm, smlh, rdlh, hot_reset;
 
-		prstn = readl(mmPSOC_GLOBAL_CONF_PRSTN);
-		ltssm = readl(mmPCIE_AUX_LTSSM_EN);
-		smlh = readl(mmPCIE_AUX_SMLH_LINK_UP);
-		rdlh = readl(mmPCIE_AUX_RDLH_LINK_UP);
-		hot_reset = readl(mmPCIE_CORE_HOT_RST);
+		prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
+		ltssm = regmap_readl(mmPCIE_AUX_LTSSM_EN);
+		smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+		rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
+		hot_reset = regmap_readl(mmPCIE_CORE_HOT_RST);
 
 		if ((hot_reset & BIT(0)) == BIT(0)) {
-			writel(0x10, mmPCIE_CORE_HOT_RST);
-			writel(1, mmPCIE_AUX_LTSSM_EN);
+			regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
+			regmap_writel(1, mmPCIE_AUX_LTSSM_EN);
 		}
 
 		if (prstn && smlh && rdlh)
@@ -106,12 +107,12 @@
 		 * till BAR is reconfigured
 		 */
 		if (!prstn && ltssm) {
-			writel(0, mmPCIE_AUX_LTSSM_EN);
+			regmap_writel(0, mmPCIE_AUX_LTSSM_EN);
 			continue;
 		}
 
 		if (prstn && !ltssm) {
-			writel(1, mmPCIE_AUX_LTSSM_EN);
+			regmap_writel(1, mmPCIE_AUX_LTSSM_EN);
 			continue;
 		}
 	}
@@ -124,8 +125,8 @@
 	int sysclk = SYS_REF_CLOCK, i;
 
 	/* timer init */
-	writel(sysclk, mmCPU_TIMESTAMP_CNTFID0);
-	writel(0x3, mmCPU_TIMESTAMP_CNTCR); /* EN | HDBG */
+	regmap_writel(sysclk, mmCPU_TIMESTAMP_CNTFID0);
+	regmap_writel(0x3, mmCPU_TIMESTAMP_CNTCR); /* EN | HDBG */
 
 	/* get begin counter value and calc end counter value */
 	asm volatile("mrs %0, cntpct_el0" : "=r" (cntpct));
@@ -145,8 +146,8 @@
 	 * DATA: mmPCIE_DB_RSV_RSV_0
 	 * STATUS: mmPCIE_DB_RSV_RSV_1
 	 */
-	writel(c, mmPCIE_DB_RSV_RSV_0);
-	writel(1, mmPCIE_DB_RSV_RSV_1);
+	regmap_writel(c, mmPCIE_DB_RSV_RSV_0);
+	regmap_writel(1, mmPCIE_DB_RSV_RSV_1);
 }
 
 u64 get_cs_pad_sel(void)
diff --git a/goya/pll.c b/goya/pll.c
index 54a03e5..269e581 100644
--- a/goya/pll.c
+++ b/goya/pll.c
@@ -3,6 +3,8 @@
 #include <linux/delay.h>
 #include <linux/io.h>
 #include <linux/types.h>
+
+#include "preboot.h"
 #include "uart.h"
 
 #define HL_PLL_NR		0x100
@@ -86,7 +88,7 @@
 {
 	int offset = out * sizeof(u32);
 
-	return readl(cfg->base + HL_PLL_DIV_SEL + offset);
+	return regmap_readl(cfg->base + HL_PLL_DIV_SEL + offset);
 }
 
 static void hl_pll_udelay(unsigned int usec)
@@ -113,19 +115,19 @@
 
 static void hl_pll_reset(struct hl_pll_cfg *cfg, bool reset)
 {
-	writel(reset, cfg->base + HL_PLL_RST);
+	regmap_writel(reset, cfg->base + HL_PLL_RST);
 }
 
 static inline int hl_pll_is_locked(struct hl_pll_cfg *cfg)
 {
-	return readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
+	return regmap_readl(cfg->base + HL_PLL_CFG) & HL_PLL_CFG_LOCK;
 }
 
 static inline void hl_pll_clear_lock_status(struct hl_pll_cfg *cfg)
 {
-	u32 reg = readl(cfg->base + HL_PLL_CFG);
+	u32 reg = regmap_readl(cfg->base + HL_PLL_CFG);
 
-	writel(reg & ~HL_PLL_CFG_LOCK, cfg->base + HL_PLL_CFG);
+	regmap_writel(reg & ~HL_PLL_CFG_LOCK, cfg->base + HL_PLL_CFG);
 }
 
 static int hl_pll_cfg(struct hl_pll_cfg *cfg)
@@ -134,11 +136,11 @@
 
 	hl_pll_reset(cfg, true);
 
-	writel(cfg->nr, cfg->base + HL_PLL_NR);
-	writel(cfg->nf, cfg->base + HL_PLL_NF);
-	writel(cfg->od, cfg->base + HL_PLL_OD);
-	writel(cfg->nb, cfg->base + HL_PLL_NB);
-	writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
+	regmap_writel(cfg->nr, cfg->base + HL_PLL_NR);
+	regmap_writel(cfg->nf, cfg->base + HL_PLL_NF);
+	regmap_writel(cfg->od, cfg->base + HL_PLL_OD);
+	regmap_writel(cfg->nb, cfg->base + HL_PLL_NB);
+	regmap_writel(0x11, cfg->base + HL_PLL_DATA_CHNG);
 
 	/* after updating the coefficients we need to wait >=5us */
 	hl_pll_udelay(HL_PLL_MAX_WAIT_TIME);
@@ -175,9 +177,9 @@
 
 		/* -1 means always bypass */
 		if (cfg->div_factor[sel] == -1 || bypass)
-			writel(HL_PLL_REF_CLK, reg);
+			regmap_writel(HL_PLL_REF_CLK, reg);
 		else
-			writel(cfg->div_factor[sel] ?
+			regmap_writel(cfg->div_factor[sel] ?
 					HL_PLL_DIV_PLL : HL_PLL_PLL_CLK, reg);
 	}
 
@@ -196,9 +198,10 @@
 
 		div_base = cfg->base + (sel * sizeof(u32));
 
-		writel(cfg->div_factor[sel], div_base + HL_PLL_DIV_FACTOR);
-		writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
-		writel(1, div_base + HL_PLL_DIV_EN);
+		regmap_writel(cfg->div_factor[sel],
+				div_base + HL_PLL_DIV_FACTOR);
+		regmap_writel(1, div_base + HL_PLL_DIV_FACTOR_CMD);
+		regmap_writel(1, div_base + HL_PLL_DIV_EN);
 	}
 
 	/* just to be on a safe side, wait for DIV init to propagate */
@@ -207,7 +210,7 @@
 
 static void hl_pll_ack_int(struct hl_pll_cfg *cfg)
 {
-	writel(1, cfg->base + HL_PLL_LOCK_INTR);
+	regmap_writel(1, cfg->base + HL_PLL_LOCK_INTR);
 }
 
 static int hl_pll_set(struct hl_pll_cfg *cfg)
